hydroflow_lang/src/graph/mod.rs

```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
-                    (Some(Color::Pull), Some(Color::Push)) => true,
+            // If SRC and DST can be in the same subgraph.
+            let can_connect = match (node_color[src], node_color[dst]) {
+                (Some(Color::Pull), Some(Color::Pull)) => true,
+                (Some(Color::Pull), Some(Color::Comp)) => true,
+                (Some(Color::Pull), Some(Color::Push)) => true,
 
-                    (Some(Color::Comp | Color::Push), Some(Color::Pull)) => false,
-                    (Some(Color::Comp | Color::Push), Some(Color::Comp)) => false,
-                    (Some(Color::Comp | Color::Push), Some(Color::Push)) => true,
+                (Some(Color::Comp | Color::Push), Some(Color::Pull)) => false,
+                (Some(Color::Comp | Color::Push), Some(Color::Comp)) => false,
+                (Some(Color::Comp | Color::Push), Some(Color::Push)) => true,
 
-                    // Handoffs are not part of subgraphs.
-                    (Some(Color::Hoff), Some(_)) => false,
-                    (Some(_), Some(Color::Hoff)) => false,
+                // Handoffs are not part of subgraphs.
+                (Some(Color::Hoff), Some(_)) => false,
+                (Some(_), Some(Color::Hoff)) => false,
 
-                    // Linear chain.
-                    (None, None) => true,
+                // Linear chain.
+                (None, None) => true,
 
-                    _some_none => unreachable!(),
-                };
-                if can_connect {
-                    node_union.union(src, dst);
-                    subgraph_edges.insert(((src, src_idx), (dst, dst_idx)));
-                    updated = true;
-                }
-            }
-            if !updated {
-                break;
+                _some_none => unreachable!(),
+            };
+            if can_connect {
+                node_union.union(src, dst);
+                subgraph_edges.insert(((src, src_idx), (dst, dst_idx)));
+                updated = true;
             }
         }
+        if !updated {
+            break;
+        }
+    }
 
-        // Copy of `self.preds` for the output.
-        let mut new_preds: SecondaryMap<GraphNodeId, OutboundEdges> = flat_graph
-            .nodes
-            .keys()
-            .map(|k| (k, Default::default()))
-            .collect();
-        // Copy of `self.succs` for the output.
-        let mut new_succs: SecondaryMap<GraphNodeId, OutboundEdges> = flat_graph
-            .nodes
-            .keys()
-            .map(|k| (k, Default::default()))
-            .collect();
-
-        // Copy over edges, inserting handoffs between subgraphs (or on subgraph self-edges) when needed.
-        for edge in iter_edges(&flat_graph.succs) {
-            let is_subgraph_edge = subgraph_edges.contains(&edge); // Internal subgraph edges are not handoffs.
-            let ((src, src_idx), (dst, dst_idx)) = edge;
+    // Copy of `self.preds` for the output.
```

Comment:
I wonder how expensive it is to work with preds and succs, and whether there is a cheaper way to do things without both. Possible add a comment for future benchmarking/perf tuning.